client/lib/cgutil/cpuset_manager_linux.go

```diff
@@ -0,0 +1,303 @@
+			reservedCpuset = reservedCpuset.Union(task.Cpuset)
+			taskCpusets[task.CgroupPath] = task
+		}
+	}
+
+	// look for reserved cpusets which we don't know about and remove
+	files, err := ioutil.ReadDir(c.reservedCpusetPath())
+	if err != nil {
+		c.logger.Error("failed to list files in reserved cgroup path during reconciliation", "path", c.reservedCpusetPath(), "error", err)
+	}
+	for _, f := range files {
+		if !f.IsDir() {
+			continue
+		}
+		path := filepath.Join(c.reservedCpusetPath(), f.Name())
+		if _, ok := taskCpusets[path]; ok {
+			continue
+		}
+		c.logger.Debug("removing reserved cpuset cgroup", "path", path)
+		err := cgroups.RemovePaths(map[string]string{"cpuset": path})
+		if err != nil {
+			c.logger.Error("removal of existing cpuset cgroup failed", "path", path, "error", err)
+		}
+	}
+
+	if err := c.setCgroupCpusetCPUs(c.sharedCpusetPath(), sharedCpuset.String()); err != nil {
+		c.logger.Error("could not write shared cpuset.cpus", "path", c.sharedCpusetPath(), "cpuset.cpus", sharedCpuset.String(), "error", err)
+	}
+	if err := c.setCgroupCpusetCPUs(c.reservedCpusetPath(), reservedCpuset.String()); err != nil {
+		c.logger.Error("could not write reserved cpuset.cpus", "path", c.reservedCpusetPath(), "cpuset.cpus", reservedCpuset.String(), "error", err)
+	}
+	for _, info := range taskCpusets {
+		if err := os.Mkdir(info.CgroupPath, 0755); err != nil && !os.IsExist(err) {
+			c.logger.Error("failed to create new cgroup path for task", "path", info.CgroupPath, "error", err)
+			info.Error = err
+			continue
+		}
+
+		// copy cpuset.mems from parent
+		_, parentMems, err := getCpusetSubsystemSettings(filepath.Dir(info.CgroupPath))
+		if err != nil {
+			c.logger.Error("failed to read parent cgroup settings for task", "path", info.CgroupPath, "error", err)
+			info.Error = err
+			continue
+		}
+		if err := fscommon.WriteFile(info.CgroupPath, "cpuset.mems", parentMems); err != nil {
+			c.logger.Error("failed to write cgroup cpuset.mems setting for task", "path", info.CgroupPath, "mems", parentMems, "error", err)
+			info.Error = err
+			continue
+		}
+		if err := c.setCgroupCpusetCPUs(info.CgroupPath, info.Cpuset.String()); err != nil {
+			c.logger.Error("failed to write cgroup cpuset.cpus settings for task", "path", info.CgroupPath, "cpus", info.Cpuset.String(), "error", err)
+			info.Error = err
+			continue
+		}
+	}
+}
+
+// setCgroupCpusetCPUs will compare an existing cpuset.cpus value with an expected value, overwriting the existing if different
+// must hold a lock on cpusetManager.mu before calling
+func (_ *cpusetManager) setCgroupCpusetCPUs(path, cpus string) error {
+	currentCpusRaw, err := fscommon.ReadFile(path, "cpuset.cpus")
+	if err != nil {
+		return err
+	}
+
+	if cpus != strings.TrimSpace(currentCpusRaw) {
+		if err := fscommon.WriteFile(path, "cpuset.cpus", cpus); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *cpusetManager) signalReconcile() {
+	c.signalCh <- struct{}{}
```

Comment:
do we need to select and ensure we aren't done since this is called from a go routine?